src/p2pLayer.ts

```diff
@@ -0,0 +1,217 @@
+                    this.socket.localAddress,
+                    this.socket.localPort
+                );
+                console.log(
+                    "   remote = %s:%s",
+                    this.socket.remoteAddress,
+                    this.socket.remotePort
+                );
+                this.sendP2pMessage(MessageType.ACK_ID);
+
+                this.socket.on("data", (data: any) => {
+                    this.onP2pMessage(data);
+                });
+                this.socket.on("end", () => {
+                    console.log("TCP disconnected");
+                });
+                this.socket.on("error", (err: any) => {
+                    console.log("Socket Error: ", JSON.stringify(err));
+                });
+                this.socket.on("close", () => {
+                    console.log("Socket Closed");
+                });
+            }
+        );
+    }
+
+    private sendP2pMessage(messageType: MessageType): void {
+        switch (messageType) {
+            case MessageType.ACK_ID: {
+                console.log("Send ACK_ID Message");
+                const nodeId = new NodeId(this.socket.localAddress, PORT);
+                const msg = new HandshakeMessage({
+                    type: "sync",
+                    version: 0,
+                    port: PORT,
+                    nodeId
+                });
+                const signedMsg = new SignedMessage(
+                    msg,
+                    this.session.getTargetNonce()
+                );
+                this.writeData(signedMsg.rlpBytes());
+                break;
+            }
+            case MessageType.REQUEST_ID: {
+                console.log("Send REQUEST_ID Message");
+                const extensionName = [
+                    "block-propagation",
+                    "parcel-propagation"
+                ];
+                let msg = new NegotiationMessage(0, 0, {
+                    type: "request",
+                    extensionName: extensionName[0],
+                    extensionVersion: [0]
+                });
+                let signedMsg = new SignedMessage(
+                    msg,
+                    this.session.getTargetNonce()
+                );
+                this.writeData(signedMsg.rlpBytes());
+
+                msg = new NegotiationMessage(0, 1, {
+                    type: "request",
+                    extensionName: extensionName[1],
+                    extensionVersion: [0]
+                });
+                signedMsg = new SignedMessage(
+                    msg,
+                    this.session.getTargetNonce()
+                );
+                this.writeData(signedMsg.rlpBytes());
+                break;
+            }
+            default:
+                throw Error("Unimplemented");
+        }
+    }
+
+    sendExtensionMessage(
+        extensionName: string,
+        extensionVersion: number,
+        data: Buffer,
+        needEncryption: Boolean
```

Comment:
`boolean`
https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html